x-pack/plugins/cases/server/connectors/cases/cases_oracle_service.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Logger, SavedObject, SavedObjectsClientContract } from '@kbn/core/server';
+import { CASE_ORACLE_SAVED_OBJECT } from '../../../common/constants';
+import { CryptoService } from './crypto_service';
+import type { OracleKey, OracleRecord, OracleRecordCreateRequest } from './types';
+
+type OracleRecordAttributes = Omit<OracleRecord, 'id' | 'version'>;
+
+export class CasesOracleService {
+  private readonly log: Logger;
+  private readonly unsecuredSavedObjectsClient: SavedObjectsClientContract;
+  private cryptoService: CryptoService;
+
+  constructor({
+    log,
+    unsecuredSavedObjectsClient,
+  }: {
+    log: Logger;
+    unsecuredSavedObjectsClient: SavedObjectsClientContract;
+  }) {
+    this.log = log;
+    this.unsecuredSavedObjectsClient = unsecuredSavedObjectsClient;
+    this.cryptoService = new CryptoService();
+  }
+
+  public getRecordId({ ruleId, spaceId, owner, grouping }: OracleKey): string {
+    if (grouping == null && ruleId == null) {
+      throw new Error('ruleID or grouping is required');
+    }
+
+    const payload = [
+      ruleId,
+      spaceId,
+      owner,
+      this.cryptoService.stringifyDeterministically(grouping),
+    ]
+      .filter(Boolean)
+      .join(':');
+
+    return this.cryptoService.getHash(payload);
+  }
+
+  public async getRecord(recordId: string): Promise<OracleRecord> {
+    this.log.debug(`Getting oracle record with ID: ${recordId}`);
+
+    const oracleRecord = await this.unsecuredSavedObjectsClient.get<OracleRecordAttributes>(
+      CASE_ORACLE_SAVED_OBJECT,
+      recordId
+    );
+
+    return this.getRecordResponse(oracleRecord);
+  }
+
+  public async createRecord(
+    recordId: string,
+    payload: OracleRecordCreateRequest
+  ): Promise<OracleRecord> {
+    const { cases, rules, grouping } = payload;
+
+    this.log.debug(`Creating oracle record with ID: ${recordId}`);
+
+    const oracleRecord = await this.unsecuredSavedObjectsClient.create<OracleRecordAttributes>(
+      CASE_ORACLE_SAVED_OBJECT,
+      {
+        counter: 1,
+        cases,
+        rules,
+        grouping,
+        createdAt: new Date().toISOString(),
+        updatedAt: null,
+      },
+      { id: recordId }
+    );
+
+    return this.getRecordResponse(oracleRecord);
+  }
+
+  public async increaseCounter(recordId: string): Promise<OracleRecord> {
+    const { id: _, version, ...record } = await this.getRecord(recordId);
+    const newCounter = record.counter + 1;
```

Comment:
will there be no hard limit for the counter? We will `put a limit on the number of total cases that can be created on an execution` but what about total cases?

Reply:
This service is responsible for interacting with the `cases-oracle` saved object. It is the consumer's responsibility to impose any limits before updating the counter.